### ISSUE TITLE
`package-comments` ignores `//nolint` directives

### ISSUE THREAD

**Issue by G-Rath** [mgechev/revive#1363]
**Describe the bug**

**To Reproduce**
1. I updated revive `go install github.com/mgechev/revive@latest`
2. I run it with no flags or config file, on the following:

```
// myfile1.go
// Package mine does something
package mine

func doSomething() {}

// myfile2.go
//nolint:dupl
package mine

func doSomethingElse() {}
```

**Expected behavior**
No lint warnings, since the package does have a comment (in `myfile1.go`), and the other comment is a directive-type comment (specifically, it's for golangci-lint)

**Logs**

```
workspace/projects-scrap/package-comments-check via 🐹 v1.24.2 took 2s
❯ ./revive .
myfile2.go:1:1: package comment should be of the form "Package mine ..."
```

**Desktop (please complete the following information):**
 - OS: Ubuntu 20.04, via WSLv2
 - Version of Go: v1.24

**Additional context**

From what I can find, it seems like directives are meant to be ignored (e.g. https://github.com/mgechev/revive/pull/989).

The situation I'm dealing with is that the `dupl` linter is flagging code that includes the `package` line meaning we have to put our `nolint` directively immediately before that (including without any newlines between it) - while it looks like I can include a `//nolint:revive` in between, the danger is that could mistakenly lead to disabling `revive` for the whole file as that'd happen if the line is merged (`//nolint:dupl,revive`) or if the `dupl` linter line is removed (e.g. because the code is changed and then the linter complains the directive is not doing anything anymore)

**Comment by alexandear** [comment]
Thanks for the detailed report! This is indeed a very niche edge case where linter directives are colliding in a subtle and unfortunate way.

There are two possible workarounds:

1. Move the `//nolint:dupl` directive below the package statement:

```go
package mine

//nolint:dupl

func doSomethingElse() {}
```

2. Exclude `myfile2.go` from `dupl` in your golangci-lint config:

```yaml
linters:
  exclusions:
    rules:
      - path: myfile2\.go
        linters:
          - dupl
```

More on exclusions here: https://golangci-lint.run/usage/false-positives/#exclude

Hope that helps!

**Comment by G-Rath** [comment]
> Add an empty line after //nolint:dupl directive:

<details>
<summary>That won't work because the `dupl` violation starts on the `package` line</summary>

```
osv-scalibr on  linting/upgrade-v2 [!?] via 🐹 v1.24.2 via  v20.11.0 took 3s
❯ make lint
go run github.com/golangci/golangci-lint/v2/cmd/golangci-lint@v2.1.6 run ./...
clients/datasource/maven_registry_cache.go:17: 17-65 lines are duplicate of `clients/datasource/pypi_registry_cache.go:17-65` (dupl)
package datasource

import (
        "time"
)

type mavenRegistryCache struct {
        Timestamp *time.Time
        Responses map[string]response // url -> response
}

// GobEncode encodes cache to bytes.
func (m *MavenRegistryAPIClient) GobEncode() ([]byte, error) {
        m.mu.Lock()
        defer m.mu.Unlock()

        if m.cacheTimestamp == nil {
                now := time.Now().UTC()
                m.cacheTimestamp = &now
        }

        cache := mavenRegistryCache{
                Timestamp: m.cacheTimestamp,
                Responses: m.responses.GetMap(),
        }

        return gobMarshal(&cache)
}

// GobDecode encodes bytes to cache.
func (m *MavenRegistryAPIClient) GobDecode(b []byte) error {
        var cache mavenRegistryCache
        if err := gobUnmarshal(b, &cache); err != nil {
                return err
        }

        if cache.Timestamp != nil && time.Since(*cache.Timestamp) >= cacheExpiry {
                // Cache expired
                return nil
        }

        m.mu.Lock()
        defer m.mu.Unlock()

        m.cacheTimestamp = cache.Timestamp
        m.responses.SetMap(cache.Responses)

        return nil
}
clients/datasource/maven_registry_cache.go:15:1: directive `//nolint:dupl` is unused for linter "dupl" (nolintlint)
//nolint:dupl
^
2 issues:
* dupl: 1
* nolintlint: 1
exit status 1
make: *** [Makefile:9: lint] Error 1
```
</details>

> Exclude myfile2.go from dupl in your golangci-lint config and remove //nolint:dupl:

While that is a workaround I've been considering, the downside is the linter won't tell us if we no longer need the exclusion meaning other duplications could be introduced there

---

Is your "wontfix" because you don't think this should be supported, or because you don't think it would possible due to how the parsing works? as I'm happy to try take a crack at addressing this, I just want to confirm first you're open to it

**Comment by chavacava** [comment]
Hi @G-Rath, could you please check if #1366 fix the problem in your specific case?

**Comment by G-Rath** [comment]
@chavacava yup that looks like it does the trick:

```
workspace/projects-scrap/package-comments-check via 🐹 v1.24.2
❯ go run github.com/mgechev/revive@latest
myfile2.go:1:1: package comment should be of the form "Package mine ..."


workspace/projects-scrap/package-comments-check via 🐹 v1.24.2
❯ go run github.com/mgechev/revive@fix-1363

```

**Comment by G-Rath** [comment]
fwiw btw this is what I'm working on where I hit this: https://github.com/google/osv-scalibr/pull/768

This is the commit where I added the `//nolint:revive`s that I'm hoping we can make not needed in future: https://github.com/google/osv-scalibr/pull/768/commits/235d2ab16efd53028e4e2d965fd7db58a6122aef

**Comment by alexandear** [comment]
> > Add an empty line after //nolint:dupl directive:
> 
> That won't work because the `dupl` violation starts on the `package` line

Try this workaround:

```go
package datasource //nolint:dupl
```

Anyway, better to simply fix `dupl` linter issues in your case.

> Is your "wontfix" because you don't think this should be supported, or because you don't think it would possible due to how the parsing works? as I'm happy to try take a crack at addressing this, I just want to confirm first you're open to it

I marked it "wontfix" initially because:

1) It's a minor edge case;
2) The workaround seemed sufficient without the need for extra complexity in fixing revive.
3) The problem is with an external linters `dupl` and golangci-lint.

But it's definitely not a hard "no" — mostly a cost/benefit judgment. That said, it looks like @chacavaca already landed a fix, which is great.
